cedar-policy-validator/src/schema.rs

```diff
@@ -1182,6 +1215,36 @@ impl ValidatorSchema {
             )
         })
     }
+
+    pub fn action_entities(&self) -> cedar_policy_core::entities::Result<Entities> {
+        // Invert the action hierarchy to get the ancestor relation expected for
+        // the `Entity` datatype instead of descendent as stored by the schema.
+        // We could store the un-inverted `memberOf` relation for each action,
+        // but I judge that the current implementation is actually less error
+        // prone, as it minimizes the threading of data structures through some
+        // complicated bits of schema construction code, and avoids computing
+        // the TC twice.
+        let mut action_ancestors: HashMap<&EntityUID, HashSet<EntityUID>> = HashMap::new();
+        for (action_euid, action_def) in &self.action_ids {
+            for descendent in &action_def.descendants {
+                action_ancestors
+                    .entry(descendent)
+                    .or_default()
+                    .insert(action_euid.clone());
+            }
+        }
+
+        Entities::from_entities(
+            self.action_ids.iter().map(|(action_id, action)| {
+                Entity::new(
+                    action_id.clone(),
+                    action.attributes.clone(),
+                    action_ancestors.remove(action_id).unwrap_or_default(),
```

Comment:
Do you expect `action_id` to end up in `action_ancestors`? I thought it was already excluded from `action_def.descendants`?

Reply:
`remove` is used here to move the ancestors set for `action_id` or of the hash map so that I can avoid a clone.